File: Tayyibat/Services/ScoringHelper.swift
import Foundation

/// إعادة حساب نسبة الالتزام محلياً بعد تعديل المستخدم لعناصر الوجبة.
/// نسخة 2: قائم على إشارات الألوان (أخضر/أصفر/أحمر). إن غاب التصنيف الصريح
/// (بيانات v1 قبل الترقية)، يُشتقّ من حكم العنصر القديم تلقائياً عبر `FoodItem.zone`.
enum ScoringHelper {
    private static func portionWeight(_ portion: String) -> Double {
        if portion.contains("صغير") { return 1 }
        if portion.contains("كبير") { return 3 }
        return 2 // متوسطة أو غير معروف
    }

    private static func zoneFactor(_ zone: FoodZone) -> Double {
        switch zone {
        case .green: return 1.0
        case .yellow: return 0.6
        case .red: return 0.0
        }
    }

    /// يحسب النتيجة من العناصر مع تطبيق سقف 50 عند وجود عنصر أحمر مرئي بثقة.
    static func recompute(items: [FoodItem]) -> Int {
        guard !items.isEmpty else { return 0 }

        var totalWeight = 0.0
        var earned = 0.0
        var hasVisibleRed = false

        for item in items {
            let w = portionWeight(item.estimatedPortion)
            totalWeight += w
            earned += w * zoneFactor(item.zone)
            if item.zone == .red, item.confidence >= 0.7 { hasVisibleRed = true }
        }

        guard totalWeight > 0 else { return 0 }
        var score = Int((earned / totalWeight * 100).rounded())
        if hasVisibleRed { score = min(score, 50) }
        return max(0, min(100, score))
    }
}

File: android/app/src/main/java/com/tayyibat/app/service/ScoringHelper.kt
package com.tayyibat.app.service

import com.tayyibat.app.data.model.EditableItem
import com.tayyibat.app.data.model.Verdict
import kotlin.math.roundToInt

/**
 * إعادة حساب نسبة الالتزام محلياً بعد تعديل المستخدم لعناصر الوجبة.
 * (التحليل الأصلي يأتي من Gemini؛ هذا يُستخدم فقط عند التعديل اليدوي.)
 */
object ScoringHelper {
    private fun portionWeight(portion: String): Double = when {
        portion.contains("صغير") -> 1.0
        portion.contains("كبير") -> 3.0
        else -> 2.0 // متوسطة أو غير معروف
    }

    private fun verdictFactor(verdict: Verdict): Double = when (verdict) {
        Verdict.TAYYIB -> 1.0
        Verdict.CONDITIONAL -> 0.5
        Verdict.KHABITH -> 0.0
    }

    /** يحسب النتيجة من العناصر مع تطبيق سقف 60 عند وجود عنصر خبيث صريح. */
    fun recompute(items: List<EditableItem>): Int {
        if (items.isEmpty()) return 0

        var totalWeight = 0.0
        var earned = 0.0
        var hasKhabith = false

        for (item in items) {
            val w = portionWeight(item.estimatedPortion)
            totalWeight += w
            earned += w * verdictFactor(item.verdict)
            if (item.verdict == Verdict.KHABITH) hasKhabith = true
        }

        if (totalWeight <= 0) return 0
        var score = (earned / totalWeight * 100).roundToInt()
        if (hasKhabith) score = minOf(score, 60)
        return score.coerceIn(0, 100)
    }
}
